application/event_processor/change_request.py

```diff
@@ -5,18 +5,29 @@
 POSTCODE_CHANGE_KEY = "postcode"
 PUBLICNAME_CHANGE_KEY = "public_name"
 WEBSITE_CHANGE_KEY = "website"
+OPENING_DATES_KEY = "opening_dates"
+OPENING_DAYS_KEY = "opening_days"
+
+DATE_FORMAT = "%Y-%m-%d"
+TIME_FORMAT = "%H:%M"
 
 
 class ChangeRequest:
-    def __init__(self, service_id, changes) -> None:
+    def __init__(self, service_id, changes=[]):
```

Comment:
Change request objects should absolutely default to empty. The only thing I've missed here is that it should be an empty dict not a list. Bit for example if you want to build a change request. 

Think about how a change request may be built. You can create the object and then add to it.
```
cr = ChangeRequest(service_id=3838)
cr.changes["phone"] = "07893726353"
cr.changes["website"] = "www.pharmacy.com"
```


allowing you to add changes in at initialisation, but defaulting to empty allows the user of the class to put in changes at the start or after creation.

Reply:
For type hinting. This may need to be something that needs to be decided on as a team what we want the codebase to be. Maybe by the tech lead. 0% type hints, 100% or some mix.

I feel 100% type hinting is very verbose and especially for __init__ functions and some others it can already be explicit what is returned without them. But then again if you do want 100% coverage of type hints for completion it may make sense. But the python purist me dislikes them as a default.

Reply:
@ed-sparkes What is your opinion on type hints?

Reply:
@SLClayton I can see the annoyance with them but as the code base will be shifted to another team in the future does it not make sense to try to make the codebase as easy to read and understand.

Reply:
Most recently i have been a full stack Typescript developer and 100% believe the benefits outweigh the additional effort and as such my vote goes for 100% type hints, agree with Jack on this one, bit of extra effort up front reaps many benefits later

Reply:
@ed-sparkes @JackPlowman Sure, consistency within a codebase trumps all else so going forward it shall be so :)

However I really have to insist on not type hinting [Python special methods](https://www.pythonlikeyoumeanit.com/Module4_OOP/Special_Methods.html). These are just redefining already built in methods with existing defined return types. For example
`__init__`
will never return anything because it is never directly called only implicitly, the 
`__str__`
function will always be used as
`str(object) `
and will always return a string. And as they are already existing built in methods, IDEs pick up their types implicitly.

Can you see my point of view for these cases?


@JackPlowman What do you think about the change request object? Are you happy with it.

Reply:
@SLClayton I can see your perspective but wherever a variable is used a change of using the incorrect data type happens. In this __init__ function how to I know if the service id is a string or integer which it is both at different points in the code. Also Sam have you used dataclasses before? Would it not be beneficial to use them instead of some of these functions? https://docs.python.org/3/library/dataclasses.html

Reply:
@JackPlowman oh yes my apologies for being unclear, for these functions I specifically meant the return type. The parameters I agree should be typed.

Reply:
@SLClayton That's great, I'm happy with that.